Adding key/value store tutorial (#355)
fermyon/developer: content/spin/kv-store.md

 Let's get started with creating and deploying your first Spin application that uses key/value storage.

## Prerequisites

First, follow [this guide](./install.md) to install Spin v0.9.0 or later. To ensure you have the correct version, you can check with this command:

<!-- @selectiveCpy -->

```bash
$ spin --version
```

## Use Spin SDK To Open And Persist Data In A Default Key/Value Store 

First, we must open a default key/value store for our Spin application. This is a special store that every environment running Spin applications will make available for their application (which when running Spin applications locally, is built using [SQLite](https://www.sqlite.org/index.html)). Then we will complete the following actions: 
- Check if key exists and retreive the corresponding value with the Spin SDK
- List all the keys with the Spin SDK

{{ tabs "sdk-type" }}

{{ startTab "Rust"}}

> 

Reviewer: Fixes lint error

```suggestion
```

title = "Using Key Value Storage with Spin Applications"
template = "spin_main"
date = "2023-02-21T00:00:00Z"
enable_shortcodes = true
[extra]

---
- [Using Key Value Storage With Spin Applications](#using-key-value-storage-with-spin-applications)
    - [Prerequisites](#prerequisites)
    - [Use Spin SDK To Open And Persist Data In A Default Key/Value Store](#use-spin-sdk-to-open-and-persist-data-in-a-default-keyvalue-store)
    - [Configure `spin.toml` To Use A Default Key/Value Store](#configure-spintoml-to-use-a-default-keyvalue-store)
    - [Build And Deploy Your Spin App](#build-and-deploy-your-spin-app)
    - [Store and Retrieve Data From Your Default Key/Value Store](#store-and-retrieve-data-from-your-default-keyvalue-store)
    - [Conclusion](#conclusion)
    - [Next Steps](#next-steps)

## Using Key Value Storage with Spin Applications

>  Spin application are best suited for event-driven, stateless workloads that have low-latency requirements. If your Spin applications needs to persist state, the data will need to be stored in an external tool or service. In [Spin v0.9.0](https://www.fermyon.com/blog/spin-v09), we've added support for key/value storage. This is a storage option for applications that need to data in the form of key/value pairs and are satistifed by a BASE consistency model. Workload examples include general value caching, session caching, counters, serialized application state.

Reviewer 2: Spin _applications_.

Reviewer 2: I think the consistency model referenced needs at least a link.

I'd be inclined to detail the consistency model section and move it to a separate section at the end.

Author: Agreed, we should go into more detail. I'm thinking it may make sense to have a separate conceptual article that walks the user through the nuts and bolts of the KV implementation, what its intended use case is, consistency model, any limitations to be aware of, etc.

Reviewer 3: Thanks, Radu, this has been reworded in the new push and Kenzie, and I have a more conceptual article in the works https://github.com/fermyon/developer/issues/364

>  Spin application are best suited for event-driven, stateless workloads that have low-latency requirements. If your Spin applications needs to persist state, the data will need to be stored in an external tool or service. In [Spin v0.9.0](https://www.fermyon.com/blog/spin-v09), we've added support for key/value storage. This is a storage option for applications that need to data in the form of key/value pairs and are satistifed by a BASE consistency model. Workload examples include general value caching, session caching, counters, serialized application state.

Reviewer 4: ```suggestion
 Spin applications are best suited for event-driven, stateless workloads that have low-latency requirements. If your Spin application needs to persist state, the data should be stored in an external tool or service. In [Spin v0.9.0](https://www.fermyon.com/blog/spin-v09), we've added support for key/value storage. This is a storage option for applications that need to data in the form of key/value pairs and are satisfied by a BASE consistency model. Workload examples include general value caching, session caching, counters, and serialized application state.
```

>  Spin application are best suited for event-driven, stateless workloads that have low-latency requirements. If your Spin applications needs to persist state, the data will need to be stored in an external tool or service. In [Spin v0.9.0](https://www.fermyon.com/blog/spin-v09), we've added support for key/value storage. This is a storage option for applications that need to data in the form of key/value pairs and are satistifed by a BASE consistency model. Workload examples include general value caching, session caching, counters, serialized application state.

Reviewer 4: Seems weird to say "best for stateless" and then say "but for state you do this". I would thread these lines together or remove the first stateless mention and say its is best used for short-lived, event-driven functions.

Reviewer 3: Thanks Kate, this has been reworded and we also have a conceptual article coming out which can dive into some of these topics such as stateless transactions/execution, ephemeral VMs etc.
https://github.com/fermyon/developer/issues/364

>  Spin users can use new a API in the Spin SDKs for persisting and retrieving non-relational data from a key/value store across multiple requests to and from the same application. Together with the updated SDKs we are introducing a built-in, _local_ key/value store available with minimal configuration for every Spin application. The default key/value store is available for the following SKDs: Go, JavaScript/TypeScript, and Rust. 

Reviewer 4: These sentences are a little long. I'd break them up more, maybe like the following:
```suggestion
 This new key/value API is available to Spin users via various language SDKs. Now spin users can persist and retrieve non-relational data from a key/value store across multiple requests to and from the same application. Along with the updated SDKs, we are introducing a built-in, _local_ key/value SQLite database available with minimal configuration for every Spin application. The default key/value store is available for the following SKDs: Go, JavaScript/TypeScript, and Rust. 
```

 Let's get started with creating and deploying your first Spin application that uses key/value storage.

## Prerequisites

First, follow [this guide](./install.md) to install Spin v0.9.0 or later. To ensure you have the correct version, you can check with this command:

<!-- @selectiveCpy -->

```bash

> spin --version

Reviewer: ```suggestion
$ spin --version
```

```

> ## Use Spin SDK To Open And Persist Data In A Default Key/Value Store 

Reviewer 4: https://titlecaseconverter.com/

> First, we must open a default key/value store for our Spin application. This is a special store that every environment running Spin applications will make available for their application (which when running Spin applications locally, is built using [SQLite](https://www.sqlite.org/index.html)). Then we will complete the following actions: 

Reviewer 4: I think this is a good opportunity to state that this store is available only to this application's components. For this reason, it may be best to start with the `Spin.toml` rather than the guest code so we can explain that it is one store per app shared by all components that are granted access

Reviewer 3: That's a great point. We now have a newly formatted section Re: "multiple requests to and from the same application".
Also added the `spin.toml` configuration file before the other application source code.

- Check if key exists and retreive the corresponding value with the Spin SDK
- List all the keys with the Spin SDK

{{ tabs "sdk-type" }}

{{ startTab "Rust"}}

> <!-- @selectiveCpy -->

Reviewer: ```suggestion
```

            (StatusCode::OK, None)
        }
        &Method::GET => {
            // Get the value associated with the request URI, or return a 404 if it's not present
            match store.get(req.uri().path()) {
                Ok(value) => (StatusCode::OK, Some(value.into())),
                Err(Error::NoSuchKey) => (StatusCode::NOT_FOUND, None),
                Err(error) => return Err(error.into()),
            }
        }
        &Method::DELETE => {
            // Delete the value associated with the request URI, if present
            store.delete(req.uri().path())?;
            (StatusCode::OK, None)
        }
        &Method::HEAD => {
            // Like GET, except do not return the value
            match store.exists(req.uri().path()) {
                Ok(true) => (StatusCode::OK, None),
                Ok(false) => (StatusCode::NOT_FOUND, None),
                Err(error) => return Err(error.into()),
            }
        }
        // No other methods are currently supported
        _ => (StatusCode::METHOD_NOT_ALLOWED, None),
    };

    Ok(http::Response::builder().status(status).body(body)?)
}

```

{{ blockEnd }}

{{ startTab "TypeScript"}}

> <!-- @selectiveCpy -->

Reviewer: ```suggestion
```

> ```typescript

Reviewer 2: The two code samples do not match here.
I would like to consider having a canonical example application that we show for this feature (it doesn't have to be the one in either examples).

Reviewer 3: Thanks @radu-matei 
I will be looking at this tomorrow for sure; I started writing/testing each of the examples to begin addressing having a canonical example and then found that other things need documenting. For example:

```bash
$ spin build  
// snip //
error[E0432]: unresolved import `spin_sdk::key_value`
could not find `key_value` in `spin_sdk`
```
Noticed users might have to do extra things like:
- `spin templates upgrade`, and/or
- update their config i.e. `go.mod` or `package.json` or `Cargo.toml`

<img width="617" alt="Screenshot 2023-02-23 at 4 18 21 pm" src="https://user-images.githubusercontent.com/9831342/220833557-2571b532-a458-4802-a5d1-2cbe2582402f.png">

<img width="631" alt="Screenshot 2023-02-23 at 4 20 12 pm" src="https://user-images.githubusercontent.com/9831342/220833801-ad296ee7-5d9c-4de1-acb5-bd3d7bc159de.png">

<img width="527" alt="Screenshot 2023-02-23 at 4 21 29 pm" src="https://user-images.githubusercontent.com/9831342/220834202-af92515b-94d0-4430-b6c3-55031f4e7532.png">

Thought I could merge this one by close of business today, but we have a ways to go yet; perhaps end of day tomorrow we can update the code, test and close this one out. :)

Reviewer 3: Karthik and Rajat and I have completed these examples and provided curl requests for `GET`, `POST`, `HEAD` and `DELETE` in the document (via the latest push to this branch).

				http.Error(w, err.Error(), http.StatusInternalServerError)
				return
			} else if !reflect.DeepEqual(value, expected) {
				http.Error(
					w,
					fmt.Sprintf("expected %v, got %v", expected, value),
					http.StatusInternalServerError,
				)
				return
			}
		}

		{
			expected := []string{"foo"}
			if value, err := key_value.GetKeys(store); err != nil {
				http.Error(w, err.Error(), http.StatusInternalServerError)
				return
			} else if !reflect.DeepEqual(value, expected) {
				http.Error(
					w,
					fmt.Sprintf("expected %v, got %v", expected, value),
					http.StatusInternalServerError,
				)
				return
			}
		}

		if err := key_value.Delete(store, "foo"); err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}

		if exists, err := key_value.Exists(store, "foo"); err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		} else if exists {
			http.Error(w, "key was not deleted as expected", http.StatusInternalServerError)
			return
		}
	})
}

func main() {}

```

{{ blockEnd }}

{{ blockEnd }}

## Configure `spin.toml` To Use A Default Key/Value Store

{{ tabs "sdk-type" }}

{{ startTab "Rust"}}

> <!-- @selectiveCpy -->

Reviewer: ```suggestion
```

```

{{ blockEnd }}

{{ startTab "TypeScript" }}

```toml

spin_version = "1"
authors = ["Fermyon Engineering <engineering@fermyon.com>"]
description = "A simple application that exercises key-value storage."
name = "spin-key-value"
trigger = {type = "http", base = "/test"}
version = "1.0.0"

[[component]]
id = "hello"
source = "target/wasm32-wasi/release/spin_key_value.wasm"
key_value_stores = ["default"]
[component.trigger]
route = "/..."
[component.build]
command = "npm run build"

```

{{ blockEnd }}

{{ startTab "TinyGo" }}

```toml

spin_version = "1"
authors = ["Fermyon Engineering <engineering@fermyon.com>"]
name = "tinygo-key-value-example"
trigger = { type = "http", base = "/" }
version = "0.1.0"

[[component]]
id = "key-value"
source = "main.wasm"
key_value_stores = ["default"]
[component.trigger]
route = "/test"
[component.build]
command = "tinygo build -target=wasi -gc=leaking -no-debug -o main.wasm main.go"

```

{{ blockEnd }}

{{ blockEnd }}

## Build And Deploy Your Spin Application

Now let's build and deploy our Spin Application locally. Run the following command to build your application: 

<!-- @selectiveCpy -->

```bash

> spin build

Reviewer: ```suggestion
$ spin build
```

```

Now run the subsequent command to deploy your application: 

<!-- @selectiveCpy -->

```bash

> 
> spin up
> 

Reviewer: ```suggestion
$ spin up
```

```

## Store and Retrieve Data From Your Default Key/Value Store

Once you have completed this minimal configuration and deployed your application, data will be persisted across requests:

<!-- @selectiveCpy -->

```bash

> 
> curl localhost:3000/hello
> curl localhost:3000/goodbye
> curl localhost:3000/hi-again
> 

Reviewer: ```suggestion
$ curl localhost:3000/hello
$ curl localhost:3000/goodbye
$ curl localhost:3000/hi-again
```

```

> You should see the following output to confirm the key/value pairs are being stored correctly:

Reviewer 4: ```suggestion
You should see the following output to confirm the key/value pairs are stored correctly:
```

> <!-- @nocopy -->

Reviewer: Are we sure that we do not want people copying these commands? I do see the same commands listed above, maybe we can combine both so as to not be repetitive.

Reviewer 3: That's a great point @karthik2804 I will update this in a new push directly to @macolso's `kvstore` branch and then resolve this conversation here in the GH UI.

```bash

> 
> $ curl localhost:3000/hello
> {"views":1,"previous_request":""}
> $ curl localhost:3000/goodbye
> {"views":2,"previous_request":"/hello"}
> $ curl localhost:3000/hi-again
> {"views":3,"previous_request":"/goodbye"}
> 

Reviewer: ```suggestion
$ curl localhost:3000/hello
{"views":1,"previous_request":""}
$ curl localhost:3000/goodbye
{"views":2,"previous_request":"/hello"}
$ curl localhost:3000/hi-again
{"views":3,"previous_request":"/goodbye"}
```

## SQLite

SQLite is an integral part of Spin's key/value API. As of Spin v0.9.0 onwards, users can easily access, a built-in, _local_ key/value SQLite database in every Spin application. This persistent storage feature is available by default and with minimal configuration. Supported language SDKs currently include Go, JavaScript/TypeScript, and Rust. With these new changes implemented.

> Spin users can now persist and retrieve non-relational data from a key/value store across multiple requests to and from the same application; written in any of these languages.

> ### Redis, PostgreSQL & SQLite

Reviewer 4: I am curious if this section overemphasizes embedded SQLite over the key/value feature. Right now SQLite is ONE host implementation of the API. This may be a good opportunity to mention that more host implementations can be added so the same spin app can target other databases such as redis.

## Creating a New Application

As previously documented, you can go ahead and [create a new Spin application from a template](https://developer.fermyon.com/spin/quickstart#creating-a-new-spin-application-from-a-template). Before you do though, please go ahead and read the [spin template](https://developer.fermyon.com/common/cli-reference#templates) and [spin new](https://developer.fermyon.com/common/cli-reference#new) sections of the Spin Command Line Interface (CLI) documentation. You will learn how to use the commands effectively and also see many handy options to [install](https://developer.fermyon.com/common/cli-reference#install-templates) and [upgrade](https://developer.fermyon.com/common/cli-reference#upgrade-templates) templates and so forth. When you are ready, go ahead and create your new application using commands similar to the ones shown below:

{{ tabs "sdk-type" }}

{{ startTab "Rust"}}

```bash
spin new http-rust my_http_rust_app
```

{{ blockEnd }}

{{ startTab "TypeScript" }}

```bash
spin new http-ts my_http_typescript_app
```

{{ blockEnd }}

{{ startTab "TinyGo" }}

```bash
spin new http-go my_http_go_app
```

{{ blockEnd }}

{{ blockEnd }}

## Configuration

> Take special note of the `key_value_stores = ["default"]` line in the `[[component]]` area of the `spin.toml` file, as shown in the next section. A newly scaffolded application will not have this line. You will need to add it.

Reviewer 4: Why do they need to add it? Good opportunity to explain per-component access to the store

Reviewer 3: Thanks @kate-goldenring 
This is a really good point, it is covered off in the new push to this branch.

Reviewer 3: Thanks Kate, updates in latest push, and new issues open in dev docs for future articles (concept article etc.)
